compiler/optimizations/deadCodeElimination.cpp

```diff
@@ -490,13 +490,97 @@ static bool isIndexVariable(Symbol* sym, CForLoop* loop) {
   return false;
 }
 
-static  CallExpr* generateGPUCall(FnSymbol* kernel,
-                                     std::vector<VarSymbol*> actuals) {
+static Symbol* extractUpperBoundFromLoop(CForLoop* loop) {
+  if(BlockStmt* bs = toBlockStmt(loop->testBlockGet())) {
+      if(CallExpr *call = toCallExpr(bs->body.head)) {
+        if(call->isPrimitive(PRIM_LESSOREQUAL)) {
+          if(SymExpr *symExpr = toSymExpr(call->get(2))) {
+            return symExpr->symbol();
+          }
+        }
+      }
+  }
+
+  INT_FATAL("Unexpected upper bound for loop identified for extraction as GPU kernel");
+  return nullptr;
+}
+
+static Symbol* extractLowerBoundFromLoop(CForLoop* loop) {
+  if(BlockStmt* bs = toBlockStmt(loop->initBlockGet())) {
+    for_alist(entry, bs->body) {
```

Comment:
For now, can we limit these bounds checks to alists that have only one statement? Otherwise, it is probably indicative of a zippered loop, but we don't want to go there yet.

And if/when we get there, we need to do something more advanced than this. (We have to match index symbols on all three parts of the loop, ideally also extracting the stride). I imagine we should do that in a way that's combined with loop checking. i.e. we should analyze the loop body *and* the header to make sure that we understand everything in there for GPU execution.

Reply:
Fixed so we use the same approach as we do in extractUpperBoundFromLoop (that is just use bs->body.head rather than iterating over the AList).